pkg/sources/adapter/solacesource/adapter.go

```diff
@@ -0,0 +1,183 @@
+	"fmt"
+	"time"
+
+	"go.uber.org/zap"
+	"pack.ag/amqp"
+
+	cloudevents "github.com/cloudevents/sdk-go/v2"
+
+	pkgadapter "knative.dev/eventing/pkg/adapter/v2"
+	"knative.dev/pkg/logging"
+
+	"github.com/triggermesh/triggermesh/pkg/apis/sources"
+	"github.com/triggermesh/triggermesh/pkg/apis/sources/v1alpha1"
+)
+
+var _ pkgadapter.Adapter = (*solacesourceAdapter)(nil)
+
+type solacesourceAdapter struct {
+	ceClient cloudevents.Client
+	logger   *zap.SugaredLogger
+	mt       *pkgadapter.MetricTag
+
+	amqpClient *amqp.Client
+	amqpOpts   []amqp.ConnOption
+	queueName  string
+}
+
+// NewAdapter satisfies pkgadapter.AdapterConstructor.
+func NewAdapter(ctx context.Context, envAcc pkgadapter.EnvConfigAccessor, ceClient cloudevents.Client) pkgadapter.Adapter {
+	logger := logging.FromContext(ctx)
+
+	mt := &pkgadapter.MetricTag{
+		ResourceGroup: sources.CloudEventsSourceResource.String(),
+		Namespace:     envAcc.GetNamespace(),
+		Name:          envAcc.GetName(),
+	}
+
+	env := envAcc.(*envAccessor)
+
+	idleTimeout := time.Duration(env.IdleTimeout) * time.Minute
+	connOption := amqp.ConnSASLAnonymous()
+
+	if env.SASLEnable {
+		connOption = amqp.ConnSASLPlain(env.Username, env.Password)
+	}
+
+	if env.TLSEnable {
+		tlsCfg := &tls.Config{}
+		if env.CA != "" {
+			addCAConfig(tlsCfg, env.CA)
+		}
+
+		if env.ClientCert != "" || env.ClientKey != "" {
+			if err := addTLSCerts(tlsCfg, env.ClientCert, env.ClientKey); err != nil {
+				logger.Panicw("Could not parse the TLS Certificates", zap.Error(err))
+			}
+		}
+
+		tlsCfg.InsecureSkipVerify = env.SkipVerify
+		connOption = amqp.ConnTLSConfig(tlsCfg)
+	}
+
+	amqpOpts := []amqp.ConnOption{
+		amqp.ConnIdleTimeout(idleTimeout),
+		connOption,
+	}
+	// Create amqp Client
+	amqpClient, err := amqp.Dial(env.URL, amqpOpts...)
+	if err != nil {
+		logger.Panicw("Could not create the amqp client", zap.Error(err))
+	}
+
+	return &solacesourceAdapter{
+		amqpClient: amqpClient,
+		amqpOpts:   amqpOpts,
+		queueName:  env.QueueName,
+
+		ceClient: ceClient,
+		logger:   logger,
+		mt:       mt,
+	}
+}
+
+func (a *solacesourceAdapter) Start(ctx context.Context) error {
+	a.logger.Info("Starting Solace Source Adapter")
+
+	// Open a session
+	session, err := a.amqpClient.NewSession()
+	if err != nil {
+		return err
+	}
+
+	// Create a receiver
+	receiver, err := session.NewReceiver(
+		amqp.LinkSourceAddress(a.queueName),
+	)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		ctx, cancel := context.WithTimeout(ctx, 1*time.Second)
+		receiver.Close(ctx)
+		cancel()
+	}()
+
+	for {
+		// Receive message
+		msg, err := receiver.Receive(ctx)
+		if err != nil {
+			a.logger.Panicw("Error receiving messages", zap.Error(err))
+		}
+
+		err = msg.Accept()
+		if err != nil {
+			a.logger.Panicw("Error accepting messages", zap.Error(err))
+
+		}
+
+		err = a.emitEvent(ctx, msg)
+		if err != nil {
+			a.logger.Panicw("Error sending event", zap.Error(err))
+		}
+
+		if ctx.Err() != nil {
+			return nil
+		}
```

Comment:
If the context was closed after the call to receive.Receive, the msg var would be empty and msg.Accept and emitEvent would most probably error.

Also Receive when the context if finished returns the ctx.Err() as an error, which means that it will go through its error check and report it as an error.
Would it make sense to move this block just after the call to Receive?